### PR TITLE
[SPARK-16240][ML] ML persistence backward compatibility for LDA

### PR DESCRIPTION
**Pull request by jkbradley**
## What changes were proposed in this pull request?

Allow Spark 2.x to load instances of LDA, LocalLDAModel, and DistributedLDAModel saved from Spark 1.6.
## How was this patch tested?

I tested this manually, saving the 3 types from 1.6 and loading them into master (2.x).  In the future, we can add generic tests for testing backwards compatibility across all ML models in SPARK-15573.

### CODE REVIEW

**Comment by jkbradley** [comment]
**To reviewers**: This code was taken and modified from [https://github.com/apache/spark/pull/14112].  @GayathriMurali should be the primary author when we merge this into master and branch-2.0

**Comment by jkbradley** [comment]
CC @hhbyyh Would you mind taking a look? thanks!

File: mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala [hunk 1]
```diff
@@ -354,6 +361,39 @@ private[clustering] trait LDAParams extends Params with HasFeaturesCol with HasM
...
+   * Equivalent to [[DefaultParamsReader.getAndSetParams()]], but handles [[LDA]] and [[LDAModel]]
+   * formats saved with Spark 1.6, which differ from the formats in Spark 2.0+.
+   *
+   * @param model    [[LDA]] or [[LDAModel]] instance.  This instance will be modified with
+   *                 [[Param]] values extracted from metadata.
+   * @param metadata Loaded model metadata
+   */
+  def getAndSetParams(model: LDAParams, metadata: Metadata): Unit = {
+    VersionUtils.majorMinorVersion(metadata.sparkVersion) match {
+      case (1, 6) =>
```
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala]: we didn't support LDA serialization in 1.5, right?

**Comment by hhbyyh** [comment]
Thanks @jkbradley. This matches what I had in mind. LGTM.

**Comment by jkbradley** [comment]
Thanks @hhbyyh !

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
@jkbradley, it looks like this is legitimately failing MiMa (not sure why it passed on the first run...):

```
[error]  * the type hierarchy of object org.apache.spark.ml.clustering.LDA is different in current version. Missing types {org.apache.spark.ml.util.DefaultParamsReadable}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("org.apache.spark.ml.clustering.LDA$")
```

**Comment by jkbradley** [comment]
@JoshRosen That is weird; MiMa passes for me locally, but I see that it shouldn't.  I added a MiMaException; this should not be a problem for users.

**Comment by jkbradley** [comment]
I'll go ahead and merge this.  Thanks @hhbyyh for reviewing it!

**Pull request closed without merging**